src/rust/dns/lib.rs

```diff
@@ -0,0 +1,174 @@
+mod ffi {
+    /// CAA record representation
+    struct CaaRecord {
+        critical: u8,
+        field: String,
+        value: String,
+    }
+    /// NAPTR record representation
+    struct NaptrRecord {
+        flags: String,
+        service: String,
+        regexp: String,
+        replacement: String,
+        order: u32,
+        preference: u32,
+    }
+    extern "Rust" {
+        fn parse_caa_record(record: &str) -> CaaRecord;
+        fn parse_naptr_record(record: &str) -> NaptrRecord;
+    }
+}
+
+/// Given a vector of strings, converts each slice to UTF-8 from HEX.
+///
+/// # Panics
+/// Will panic if any substring is not a valid hex.
+#[must_use]
+pub fn decode_hex(s: &[&str]) -> Vec<String> {
+    s.iter()
+        .map(|s| {
+            u16::from_str_radix(s, 16)
+                .ok()
+                .map(|b| String::from_utf16(&[b]).unwrap())
+                .unwrap()
+        })
+        .collect()
+}
+
+/// Parses an unknown RR format returned from Cloudflare DNS.
+/// Specification is available at
+/// `<https://datatracker.ietf.org/doc/html/rfc3597>`
+///
+/// The format of the record is as follows:
+///   \# <length-in-bytes> <bytes-in-hex>
+///   \\# 15 00 05 69 73 73 75 65 70 6b 69 2e 67 6f 6f 67
+///       |  |  |  |
+///       |  |  |  - Starting point of the actual data
+///       |  |  - Length of the field.
+///       |  - Number representation of "`is_critical`"
+///       - Length of the data
+///
+/// Note: Field can be "issuewild", "issue" or "iodef".
+///
+/// ```
+/// let record = parse_caa_record("\\# 15 00 05 69 73 73 75 65 70 6b 69 2e 67 6f 6f 67");
+/// assert_eq!(record.critical, false);
+/// assert_eq!(record.field, "issue")
+/// assert_eq!(record.value, "pki.goog")
+/// ```
+fn parse_caa_record(record: &str) -> ffi::CaaRecord {
+    // Let's remove "\\#" and the length of data from the beginning of the record
+    let data = record.split_ascii_whitespace().collect::<Vec<_>>()[2..].to_vec();
+    let critical = data[0].parse::<u8>().unwrap();
+    let prefix_length = data[1].parse::<usize>().unwrap();
+
+    let field = decode_hex(&data[2..prefix_length + 2]).join("");
+    let value = decode_hex(&data[(prefix_length + 2)..]).join("");
+
+    // Field can be "issuewild", "issue" or "iodef"
+    assert!(
+        field == "issuewild" || field == "issue" || field == "iodef",
+        "received unsupported field {field}",
+    );
+
+    ffi::CaaRecord {
+        critical,
+        field,
+        value,
+    }
+}
+
+/// Parses an unknown RR format returned from Cloudflare DNS.
+/// Specification is available at
+/// `<https://datatracker.ietf.org/doc/html/rfc3597>`
+///
+/// The format of the record is as follows:
+/// \# 37 15 b3 08 ae 01 73 0a 6d 79 2d 73 65 72 76 69 63 65 06 72 65 67 65 78 70 0b 72 65 70 6c 61 63 65 6d 65 6e 74 00
+///       |--|  |--|  |  |  |  |--------------------------|  |  |--------------|  |  |--------------------------------|
```

Comment:
Nice :)

Is `- Replacement length` wrongly positioned?

Reply:
Yes! Nice catch.